torch/utils/bottleneck/__main__.py

```diff
@@ -63,86 +72,103 @@ def compiled_with_cuda():
-
-    return '\n'.join(result)
-
-
-def set_env_variables(gpu_device):
-    # Override CUDA_LAUNCH_BLOCKING by default for more accurate profiling.
-    os.environ['CUDA_LAUNCH_BLOCKING'] = '1'
-
-    # Only profile on one GPU for simplicity
-    os.environ['CUDA_VISIBLE_DEVICES'] = '{}'.format(gpu_device)
-
+        cuda = 'CUDA unavailable'
+
+    pip_version, pip_list_output = check_pip_packages()
+    if pip_list_output is None:
+        pip_list_output = 'Unable to fetch'
+
+    result = {
+        'debug_str': debug_str,
+        'pytorch_version': torch.__version__,
+        'cuda_compiled': compiled_with_cuda(),
+        'py_version': '{}.{}'.format(sys.version_info[0], sys.version_info[1]),
+        'cuda_runtime': cuda_avail,
+        'pip_version': pip_version,
+        'pip_list_output': pip_list_output,
+    }
 
-def get_env_variables_description(gpu_device, prefix=' '):
-    if torch.cuda.is_available():
-        return '{}with environment variables CUDA_LAUNCH_BLOCKING=1 CUDA_VISIBLE_DEVICES={}'.format(
-            prefix, gpu_device)
-    else:
-        return ''
+    return env_summary.format(**result)
 
 
-def run_cprofile(code, globs, gpu_device=0):
-    set_env_variables(gpu_device)
-    env_str = get_env_variables_description(gpu_device)
-    print('Running your script with cProfile{}...'.format(env_str))
+def run_cprofile(code, globs, launch_blocking=False):
+    print('Running your script with cProfile')
     prof = cProfile.Profile()
     prof.enable()
     exec(code, globs, None)
     prof.disable()
     return prof
 
 
-def print_line(width=80):
-    print('-' * width)
+cprof_summary = """
+--------------------------------------------------------------------------------
+  cProfile output
+--------------------------------------------------------------------------------
+""".strip()
+
+
+def print_cprofile_summary(prof, sortby='tottime', topk=15):
+    result = {}
 
+    print(cprof_summary.format(**result))
 
-def print_cprofile_summary(prof, gpu_device=0, sortby='tottime', topk=15):
-    print_line()
-    env_str = get_env_variables_description(gpu_device, '\n')
-    print('cProfile output{}'.format(env_str))
-    print_line()
     cprofile_stats = pstats.Stats(prof).sort_stats(sortby)
     cprofile_stats.print_stats(topk)
 
 
-def run_autograd_prof(code, globs, gpu_device=0):
-    set_env_variables(gpu_device)
-    env_str = get_env_variables_description(gpu_device)
-    print('Running your script with the autograd profiler{}...'.format(env_str))
-    with profiler.profile() as prof:
-        exec(code, globs, None)
-    return prof
+def run_autograd_prof(code, globs):
+    def run_prof(use_cuda=False):
+        with profiler.profile(use_cuda=use_cuda) as prof:
+            exec(code, globs, None)
+        return prof
+
+    print('Running your script with the autograd profiler...')
+    result = [run_prof(use_cuda=False)]
+    if torch.cuda.is_available():
+        result.append(run_prof(use_cuda=True))
```

Comment:
Oh I've been thinking of running the user script twice - once with CUDA tracing enabled, and once with it disabled. It's quite important, because CUDA kernel profiling adds a ton of overhead to, otherwise pretty much free, profiler.

Reply:
Do you mean with the `profiler.emit_nvtx`?

Reply:
No, `torch.cuda.profiler.profile(use_cuda=True)` is as expensive as `emit_nvtx`. The benefit is that you don't need to emit, stop the process (so nvprof flushes the output to disk), open a repl, load and inspect the results there

Reply:
Understood, thank you for the explanation. Right now in its current iteration the script will run the user script twice, once with `.profile(use_cuda=True)` and once with `.profile(use_cuda=False), if CUDA is available and display both results if the total execution times between the two differ by more than 5%. Is this something along the lines of what you were suggesting?

Reply:
Oh I must have missed that. That's great!